Fix type parsing issues in ILLink and ILC

sbomer opened the pull request:
- Fixes https://github.com/dotnet/runtime/issues/98955

   We will now produce a warning when a non-assembly-qualified type flows into a string location annotated with DynamicallyAccessedMembers, and we don't try to look up or mark the type (since we don't know which assemblies will be searched at runtime by the Type.GetType call).

- Fixes https://github.com/dotnet/runtime/issues/103906

   The ILLink intrinsic handling for `Type.GetType` will now look in corelib for generic arguments, matching native AOT.

This replaces the existing warning IL2105. I left that one alone instead of repurposing it, because we already documented IL2105 and older versions of ILLink will produce it. Best to avoid any confusion about them.

sbomer committed:
Fix type name parsing issues in ILLink and ILC

sbomer committed:
Add test for generic over corelib type

sbomer committed:
Fix tests

- Revert change to needsAssemblyName for types in attribute XML
- Update SharedStrings.resx to match renamed DiagnosticId

sbomer committed:
Mark IL2105 as unused

jkotas commented on src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs:
```diff
@@ -52,8 +57,28 @@ public static class CustomAttributeTypeNameParser
+                    return false;
+                }
+
+                if (typeName.IsConstructedGenericType)
```
This does not seem to handle arrays/byrefs/pointers of generic types with unqualified argument.

jkotas replied:
I meant type name like `G`1[[MyType]][], MyAssembly`. I still think that type names like that are not handled correctly here.

(The tests you have added are for type names like `G`1[[MyType[]]], MyAssembly`. I agree that those are handled correctly.)

sbomer committed:
PR feedback

- Test array/byref/pointer that's not assembly-qualified

Also:
- Fix ILCompiler.Trimming.Tests build
- Clean up usings

sbomer committed:
PR feedback

- Fix check for array/pointer/byref types, update tests

sbomer committed:
Use System.Runtime as system module in library mode

sbomer committed:
Fix ApiCompat

sbomer committed:
Various fixes

- Fix ILCompiler.Trimming.Tests build
- Replace IL2105 suppressions with IL2122
- Fix trimming tests that were not using fully-qualified type names

eerhardt commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/TrimmingTests/DebuggerTypeProxyAttributeTests.cs:
```diff
@@ -53,7 +53,7 @@ public DebuggerProxy(MyClass instance)
     }
 
-    [DebuggerTypeProxy("Program+MyClassWithProxyStringProxy")]
+    [DebuggerTypeProxy("Program+MyClassWithProxyStringProxy, project")]
```
I think I'm fine with this behavior. I searched a few places that use Type.GetType currently, and they all appear to be using assembly qualified names. If we see new warnings, we should fix them.

sbomer committed:
Fix another test

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Common/Utilities/CustomAttributeTypeNameParser.cs:
```diff
@@ -52,8 +57,33 @@ public static class CustomAttributeTypeNameParser
-            typeWasNotFoundInAssemblyNorBaseLibrary = resolver._typeWasNotFoundInAssemblyNorBaseLibrary;
             return type;
+
+            static bool IsFullyQualified (TypeName typeName)
```
We don't do this around here

```suggestion
            static bool IsFullyQualified(TypeName typeName)
```

MichalStrehovsky commented on src/tools/illink/src/linker/Linker.Steps/RootAssemblyInputStep.cs:
```diff
@@ -86,6 +86,9 @@ protected override void Process ()
 				// No metadata trimming
 				Context.MetadataTrimming = MetadataTrimming.None;
+
+				// Library mode builds against System.Runtime ref assembly
+				Context.SystemModuleName = "System.Runtime";
```
We should know what the core assembly is (it's the one that defines System.Object). Why is this extra configuration needed? I think I didn't fully understand https://github.com/dotnet/runtime/pull/104060#discussion_r1655659027.

sbomer committed:
PR feedback

- Resolve object to find core library

sbomer committed:
Merge branch 'main' into fixTypeParsing

jonathanpeppers commented:
@sbomer is it possible IL2122 might still appear even if `SuppressTrimAnalysisWarnings=true` / `--notrimwarn` was passed in?

![image](https://github.com/user-attachments/assets/d2553924-71bc-4216-873f-d83c71c21d7e)

We noticed this here:
* https://github.com/dotnet/android/pull/9108
